### PR TITLE
[WFLY-12457] Single action timer is not triggered automatically after a DB outage, requires server restart

### PR DESCRIPTION
**Pull request by elguardian**
added logic when in memory timer are out of sync with the store

https://issues.jboss.org/browse/WFLY-12457

### CODE REVIEW

**Comment by bstansberry** [comment]
Hi @elguardian  -- Please file a WFLY JIRA for this and use it in the commit message and PR title. Thanks!

**Comment by elguardian** [comment]
@bstansberry done, who should I ask for review ?
Cannot access the jdk 11 errors.

**Comment by bstansberry** [comment]
@chengfang or @tadamski Please review when you get a chance.

@elguardian The JDK 11 errors look unrelated.

File: ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/TimerPersistence.java [hunk 1]
```diff
@@ -96,6 +96,12 @@
          */
         void timerAdded(TimerImpl timer);
 
+        /**
+         * Invoked when a timer needs to be sync with the underlying store
+         * @param timer The timer
+         */
```
Comment (chengfang) [inline ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/TimerPersistence.java]: document both params.

File: ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java [hunk 2]
```diff
@@ -839,6 +840,23 @@ public void run() {
...
+                                    } else {
+                                        final Holder holder = timerFromResult(resultSet, listener.getTimerService());
+                                        TimerImpl oldTimer = listener.getTimerService().getTimer(id);
+                                        // if it is already in memory but it is not in sync we have a problem
+                                        // remove and add -> the probable cause is db glitch
+                                        EnumSet<TimerState> valid = EnumSet.of(TimerState.IN_TIMEOUT, TimerState.RETRY_TIMEOUT, TimerState.CREATED, TimerState.ACTIVE);
+                                        boolean validDBTimer = valid.contains(holder.timer.getState());
+                                        boolean validMemoryTimer = !valid.contains(oldTimer.getState());
+                                        // if timers memory - db are in non intersect substes of valid/invalid states. we put them in sync
+                                        if (holder != null && validMemoryTimer && validDBTimer) {
```
Comment (chengfang) [inline ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java]: move holder != null check earlier before its first access.

**Review by chengfang — COMMENTED** [review]
Looks good other than a few minor points noted inline.

File: ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/TimerPersistence.java (hunk 1 above)
Comment (elguardian) [reply ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/TimerPersistence.java]: done

File: ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java (hunk 2 above)
Comment (elguardian) [reply ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java]: done

**Comment by elguardian** [comment]
@chengfang both checks are done. let me know if there is anything to make this move forward

**Review by chengfang — APPROVED** [review]

**Comment by bstansberry** [comment]
@elguardian I added a commit fixing the checkstyle problem.

**Review by bstansberry — APPROVED** [review]

**Comment by elguardian** [comment]
thx @bstansberry

**Pull request merged**